parquet/variants/primitive.go

```diff
@@ -0,0 +1,678 @@
+// Timezone information can be determined from a time.Time struct. Otherwise, by default, timestamps will be written with
+// local timezone set.
+func marshalPrimitive(v any, w io.Writer, opts ...MarshalOpts) (int, error) {
+	var allOpts MarshalOpts
+	for _, o := range opts {
+		allOpts |= o
+	}
+	switch val := v.(type) {
+	case bool:
+		return marshalBoolean(val, w), nil
+	case int:
+		return marshalInt(int64(val), w), nil
+	case int8:
+		return marshalInt(int64(val), w), nil
+	case int16:
+		return marshalInt(int64(val), w), nil
+	case int32:
+		return marshalInt(int64(val), w), nil
+	case int64:
+		if allOpts&MarshalAsTime != 0 {
+			encodeTimestamp(val, allOpts&MarshalTimeNanos != 0, allOpts&MarshalTimeNTZ != 0, w)
+		}
+		return marshalInt(val, w), nil
+	case float32:
+		return marshalFloat(val, w), nil
+	case float64:
+		return marshalDouble(val, w), nil
+	case string:
+		if allOpts&MarshalAsUUID != 0 {
+			return marshalUUID([]byte(val), w), nil
+		}
+		return marshalString(val, w), nil
+	case []byte:
+		if allOpts&MarshalAsUUID != 0 {
+			return marshalUUID([]byte(val), w), nil
+		}
+		return marshalBinary(val, w), nil
+	case time.Time:
+		if allOpts&MarshalAsDate != 0 {
+			return marshalDate(val, w), nil
+		}
+		return marshalTimestamp(val, allOpts&MarshalTimeNanos != 0, w), nil
+	}
+	if v == nil {
+		return marshalNull(w), nil
+	}
+	return -1, fmt.Errorf("unsupported primitive type")
+}
+
+// unmarshals a primitive (or a short string) into dest. dest must be a non-nil pointer to variable that is
+// compatible with the Variant value to decode. Some conversions can take place:
+//   - Integer values: Higher widths can be decoded into smaller widths so long as they don't overflow. Also,
+//     integral values can be decoded into floats (also so long as they don't overflow).
+//   - Time/timestamps: Can be decoded into either int64 or time.Time, the latter of which will carry time zone information
+//   - Strings and binary: Can be decoded into either string or []byte
+//
+// If the Variant primitive is of the Null type, dest will be set to its zero value.
+func unmarshalPrimitive(raw []byte, offset int, destPtr reflect.Value) error {
+	dest := destPtr.Elem()
+	kind := dest.Kind()
+
+	if err := checkBounds(raw, offset, offset); err != nil {
+		return err
+	}
+
+	prim, err := primitiveFromHeader(raw[offset])
+	if err != nil {
+		return err
+	}
+
+	switch prim {
+	case primitiveNull:
+		dest.Set(reflect.Zero(dest.Type()))
+	case primitiveTrue, primitiveFalse:
+		if kind != reflect.Bool && kind != reflect.Interface {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+		dest.Set(reflect.ValueOf(prim == primitiveTrue))
+	case primitiveInt8, primitiveInt16, primitiveInt32, primitiveInt64:
+		iv, err := decodeIntPhysical(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(iv))
```

Comment:
I don't understand why you're checking and using `reflect.Interface` here. Can you explain? For reference, `reflect.Interface` just means the type is *an* interface, it doesn't mean it is the empty interface (i.e. `interface{}` or `any`)